src/python/bow/bow.py

```diff
@@ -0,0 +1,224 @@
+import numpy as np
+import cv2
+import argparse
+from pathlib import Path
+import matplotlib.pyplot as plt
+
+from sklearn import preprocessing
+from sklearn.cluster import KMeans
+from sklearn.neighbors import KDTree
+
+kDefaultWidth = 640  # px
+kDefaultClusterSize = 400
+
+
+def listImagesInFolder(folderPath):
+    trainImageFiles = list(folderPath.glob("*.jpg"))
+    trainImageFiles.extend(list(folderPath.glob("*.png")))
+    return trainImageFiles
+
+
+def rescaleImageIfNeeded(image):
+    height, width = image.shape
+    if width > kDefaultWidth:
+        newHeight = (height * kDefaultWidth) / width
+        image = cv2.resize(image, (kDefaultWidth, int(newHeight)))
+        print("Resized image from", height, width, "to", image.shape)
+    return image
+
+
+def extractSiftsFromImage(imageFile):
+    """Extracts SIFT features from an image
+
+    Args:
+        imageFile (Path): path to the image file
+
+    Returns:
+        list(list(int)): array of descriptors NxD
+    """
+    # Extracts features from an image
+    image = cv2.imread(imageFile.as_posix(), cv2.IMREAD_GRAYSCALE)
+    image = rescaleImageIfNeeded(image)
+    sift = cv2.SIFT_create()
+    keypoints, descriptors = sift.detectAndCompute(image, None)
+    return descriptors
+
+
+def computeIDF(descriptorsPerImage, clusters):
+    """Compute inverse document frequence (IDF). Here means in how many images does the word occur.
```

Comment:
> Here means in how many images does the word occur.

nit: This reads a bit strange. What did you mean to say here?

Reply:
I meant the IDF (which is inverse document frequency) , here stands for how many images (not documents) does the word occur in.